lstests/core/service.py

```diff
@@ -0,0 +1,1526 @@
+
+    def is_node_ready(self, node_id, **call_kwargs):
+        return "error" not in self.get_cluster_info(**call_kwargs).response_dict["result"][node_id][1]
+
+    def get_head_block_number(self, node_id=0, **call_kwargs):
+        """Get head block number by node id."""
+        return self.get_cluster_info(**call_kwargs).response_dict["result"][node_id][1]["head_block_num"]
+
+    def get_head_block_producer(self, node_id=0, **call_kwargs):
+        """Get head block producer by node id."""
+        return self.get_cluster_info(**call_kwargs).response_dict["result"][node_id][1]["head_block_producer"]
+
+    def get_running_nodes(self, **call_kwargs):
+        cluster_result = self.get_cluster_info(**call_kwargs).response_dict["result"]
+        count = 0
+        for node_result in cluster_result:
+            if "head_block_id" in node_result[1]:
+                count += 1
+        return count
+
+    def get_log(self, node_id=0, length=10000, filename="stderr_0.txt", **call_kwargs):
+        log = ""
+        offset = 0
+        while True:
+            response = self.get_log_data(node_id=node_id, offset=offset, length=length, filename=filename, **call_kwargs).response_dict
+            log += base64.b64decode(response["data"]).decode("utf-8")
+            if response["offset"] + length > response["filesize"]:
+                break
+            offset += length
+        return log
+
+    def wait_nodes_ready(self, node_list:list=None, retry=100, sleep=1, dont_raise=False, level="debug", sublevel="trace"):
+        if node_list is None: node_list = list(range(self.node_count))
+        self.print_header("wait for nodes to get ready", level=level)
+        max_wait_time = retry * sleep
+        while True:
+            result = self.get_cluster_info(level=sublevel).response_dict["result"]
+            error_node_list = [x[0] for x in result if x[0] in node_list and "error" in x[1]]
+            error_node_count = len(error_node_list)
+            if error_node_count == 0:
+                self.log(f"All {len(node_list)} nodes are ready.", level=level)
+                return True
+            if retry > 0:
+                self.log(f"Nodes that are not ready: {error_node_list}. "
+                         f"{retry} {helper.plural(['retry', 'retries'], retry)} remain. Sleep for {sleep}s.", level=level)
+                time.sleep(sleep)
+                retry -= 1
+            else:
+                msg = f"After waiting for {max_wait_time}s, there still are nodes that are not ready: {error_node_list}."
+                self.error(msg)
+                if not dont_raise:
+                    raise BlockchainError(msg)
+                return False
+
+    def wait_get_block(self, block_num, node_id=0, retry=10, dont_raise=False, level="debug", sublevel="trace"):
+        """Get block information by block num. If that block has not been produced, wait for it."""
+        for __ in range(10):
```

Comment:
Thanks for the observation. Some people argue that in Python single underscore has a dedicated meaning -- representing the last result from IDLE like
```
>>> 1 + 7
>>> 8
>>> print(_)
>>> 8
```
Because of this the double underscore may better serve as the dummy variable to be discarded. Personally, I was dubious when I put down these double underscores as the argument sounds too pedantic. Since you have the question here, I will change them back to a single underscore. Thank you.